File: object-oriented-programming/virtual-world-java/src/Game/Model/Plants/Plant.java
package Game.Model.Plants;

import Game.Model.Organism;
import Game.Model.Species;
import Game.Model.World;

import java.util.Random;

public abstract class Plant extends Organism {
    Random random = new Random();
    protected final int sowAttempts;

    public Plant(Species species, int x, int y, int strength, World world, int attempts){
        super(species, x, y, strength, 0, world, 0, true);
        this.sowAttempts = attempts;
    }

    public void action(){
        int[] dy = {1, 0, -1, 0};
        int[] dx = {0, 1, 0, -1};
        for(int i = 0; i < sowAttempts; i++){
            int direction = random.nextInt(4);
            int newY = y + dy[direction];
            int newX = x + dx[direction];
            if(world.inBounds(newY, newX) && world.getOrganism(newY, newX) == null && random.nextInt(2) % 2 == 0){
                sow(newY, newX);
                return;
            }
        }
    }

    public void collision(Organism other){
        world.report(other.getSpecies().displayName() + " ate " + species.displayName());
        markDead();
        world.setOrganism(null, y, x);
        other.moveOrganism(y, x);
    }

    public abstract Plant sow(int y, int x);
}

File: object-oriented-programming/virtual-world-python/plants/plant.py
from abc import abstractmethod
from organism import Organism
import random

class Plant(Organism) :

    def __init__(self, species, x, y, strength, world, attempts):
        super().__init__(species, x, y, strength, 0, world, True, 0)
        self._sow_attempts = attempts

    
    def action(self) :
        dy = [1, 0, -1, 0]
        dx = [0, 1, 0, -1]
        for i in range(4) :
            direction = random.randint(0, 3)
            newY = self.y + dy[direction]
            newX = self.x + dx[direction]
            if(self._world.in_bounds(newX, newY) and self._world.get_organism(newX, newY) == None and random.randint(0, 1) % 2 == 0) :
                self.sow(newX, newY)
                return
    
    def collision(self, other) :
        self._world.report(f"{other.species.display_name()} ate {self._species.display_name()}!")
        self.mark_dead()
        self._world.set_organism(None, self.x, self.y)
        other.move_organism(self.x, self.y)

    @abstractmethod
    def sow(x, y) :
        pass
    